### PR TITLE
Fix expected SFT file names when duration is not multiple of Tsft

### LINKED ISSUE #202: Wrong expected SFT filenames when duration is not multiple of Tsft (e.g. for Tsft=1024s in many examples)
The following happens when `Tsft` is set to `1024` instead of `1800` in the `twoF_cumulative` example.
```
(pyfstat) [rodrigo@spinning-quadrupole other_examples]$ python PyFstat_example_twoF_cumulative.py
11:05 INFO    : Creating BinaryModulatedWriter object...
11:05 INFO    : Generating SFTs with fmin=28.0, Band=4
11:05 INFO    : Checking if we can re-use injection config file...
11:05 INFO    : ...no config file PyFstat_example_data/PyFstat_example_twoF_cumulative/PyFstat_example_twoF_cumulative.cff found.
11:05 INFO    : Writing config file: PyFstat_example_data/PyFstat_example_twoF_cumulative/PyFstat_example_twoF_cumulative.cff
11:05 INFO    : Checking if we can re-use existing SFT data file(s)...
11:05 INFO    : ...no SFT file matching 'PyFstat_example_data/PyFstat_example_twoF_cumulative/H-843_H1_1024SFT_PyFstat_example_twoF_cumulative-1000000000-864000.sft' found. Will create new SFT file(s).
11:05 INFO    : Now executing: lalapps_Makefakedata_v5 --outSingleSFT=TRUE --outSFTdir="PyFstat_example_data/PyFstat_example_twoF_cumulative" --outLabel="PyFstat_example_twoF_cumulative" --IFOs="H1","L1" --sqrtSX="1e-23" --startTime=1000000000 --duration=864000 --fmin=28 --Band=4 --Tsft=1024 --injectionSources="PyFstat_example_data/PyFstat_example_twoF_cumulative/PyFstat_example_twoF_cumulative.cff" --ephemEarth="/home/rodrigo/gravity/venv/pyfstat/share/lalpulsar/earth00-40-DE405.dat.gz" --ephemSun="/home/rodrigo/gravity/venv/pyfstat/share/lalpulsar/sun00-40-DE405.dat.gz"
11:05 INFO    : Successfully wrote SFTs to: PyFstat_example_data/PyFstat_example_twoF_cumulative/H-843_H1_1024SFT_PyFstat_example_twoF_cumulative-1000000000-864000.sft;PyFstat_example_data/PyFstat_example_twoF_cumulative/L-843_L1_1024SFT_PyFstat_example_twoF_cumulative-1000000000-864000.sft
11:05 INFO    : Now executing: lalapps_PredictFstat --h0=9.999999999999999e-26 --cosi=1 --psi=0.39269908169872414 --Alpha=1.4596048908088417 --Delta=0.38422376285103965 --DataFiles='PyFstat_example_data/PyFstat_example_twoF_cumulative/H-843_H1_1024SFT_PyFstat_example_twoF_cumulative-1000000000-864000.sft;PyFstat_example_data/PyFstat_example_twoF_cumulative/L-843_L1_1024SFT_PyFstat_example_twoF_cumulative-1000000000-864000.sft' --minStartTime=1000000000 --maxStartTime=1000864000 --assumeSqrtSX=1e-23 --outputFstat=PyFstat_example_data/PyFstat_example_twoF_cumulative/PyFstat_example_twoF_cumulative.tmp --ephemEarth='/home/rodrigo/gravity/venv/pyfstat/share/lalpulsar/earth00-40-DE405.dat.gz' --ephemSun='/home/rodrigo/gravity/venv/pyfstat/share/lalpulsar/sun00-40-DE405.dat.gz'
11:05 ERROR   : Execution failed: Command 'lalapps_PredictFstat --h0=9.999999999999999e-26 --cosi=1 --psi=0.39269908169872414 --Alpha=1.4596048908088417 --Delta=0.38422376285103965 --DataFiles='PyFstat_example_data/PyFstat_example_twoF_cumulative/H-843_H1_1024SFT_PyFstat_example_twoF_cumulative-1000000000-864000.sft;PyFstat_example_data/PyFstat_example_twoF_cumulative/L-843_L1_1024SFT_PyFstat_example_twoF_cumulative-1000000000-864000.sft' --minStartTime=1000000000 --maxStartTime=1000864000 --assumeSqrtSX=1e-23 --outputFstat=PyFstat_example_data/PyFstat_example_twoF_cumulative/PyFstat_example_twoF_cumulative.tmp --ephemEarth='/home/rodrigo/gravity/venv/pyfstat/share/lalpulsar/earth00-40-DE405.dat.gz' --ephemSun='/home/rodrigo/gravity/venv/pyfstat/share/lalpulsar/sun00-40-DE405.dat.gz'' returned non-zero exit status 1.
11:05 ERROR   : ERROR: Failed to open matched file 'PyFstat_example_data/PyFstat_example_twoF_cumulative/H-843_H1_1024SFT_PyFstat_example_twoF_cumulative-1000000000-864000.sft'

XLAL Error - XLALSFTdataFind (SFTfileIO.c:277): I/O error
XLAL Error - InitPFS (PredictFstat.c:392): Check failed: (catalog = XLALSFTdataFind ( uvar->DataFiles, &constraints )) != ((void *)0)
XLAL Error - InitPFS (PredictFstat.c:392): Internal function call failed: I/O error
XLAL Error - main (PredictFstat.c:156): Check failed: InitPFS ( &GV, &uvar ) == XLAL_SUCCESS
XLAL Error - main (PredictFstat.c:156): Internal function call failed: I/O error

Traceback (most recent call last):
  File "PyFstat_example_twoF_cumulative.py", line 59, in <module>
    twoF = data.predict_fstat()
  File "/home/rodrigo/gravity/git/PyFstat/pyfstat/make_sfts.py", line 589, in predict_fstat
    twoF_expected, twoF_sigma = helper_functions.predict_fstat(
  File "/home/rodrigo/gravity/git/PyFstat/pyfstat/helper_functions.py", line 883, in predict_fstat
    run_commandline(cl_pfs)
  File "/home/rodrigo/gravity/git/PyFstat/pyfstat/helper_functions.py", line 303, in run_commandline
    out = subprocess.check_output(
  File "/usr/lib/python3.8/subprocess.py", line 411, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.8/subprocess.py", line 512, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command 'lalapps_PredictFstat --h0=9.999999999999999e-26 --cosi=1 --psi=0.39269908169872414 --Alpha=1.4596048908088417 --Delta=0.38422376285103965 --DataFiles='PyFstat_example_data/PyFstat_example_twoF_cumulative/H-843_H1_1024SFT_PyFstat_example_twoF_cumulative-1000000000-864000.sft;PyFstat_example_data/PyFstat_example_twoF_cumulative/L-843_L1_1024SFT_PyFstat_example_twoF_cumulative-1000000000-864000.sft' --minStartTime=1000000000 --maxStartTime=1000864000 --assumeSqrtSX=1e-23 --outputFstat=PyFstat_example_data/PyFstat_example_twoF_cumulative/PyFstat_example_twoF_cumulative.tmp --ephemEarth='/home/rodrigo/gravity/venv/pyfstat/share/lalpulsar/earth00-40-DE405.dat.gz' --ephemSun='/home/rodrigo/gravity/venv/pyfstat/share/lalpulsar/sun00-40-DE405.dat.gz'' returned non-zero exit status 1.
```

### PR DESCRIPTION
**Pull request by dbkeitel**
I think this is sufficient to fix #202. The alternative `_get_setup_from_noiseSFTs()` case should be fine already, since there we are just taking the SFT file names as we've been given them, but I haven't dug explicitly into it.

- [ ] TODO: check MFDv5 code/documentation to confirm we now actually _always_ match its behaviour, and I haven't left out weird corner cases.
- [x] TODO: add a test case

### CODE REVIEW

**Comment by dbkeitel** [comment]
I've now added a new child class of `TestWriter` that repeats all standard tests but with `Tsft=1024`.

Two of its functions failed for reasons similar to what was fixed here, already fixed.

One more (`test_data_with_gaps`) is still failing, there's a trivial workaround commented out, but it would be better to check if the `noiseSFTs` case can be made smarter to actually extract the Tsft properly.

**Comment by dbkeitel** [comment]
> One more (test_data_with_gaps) is still failing, there's a trivial workaround commented out, but it would be better to check if the noiseSFTs case can be made smarter to actually extract the Tsft properly.

Fixed properly. Now overwriting the `self.Tsft` from what's inside the `noiseSFTs`, and as a bonus throwing a warning if these are different.

**Comment by dbkeitel** [comment]
>     * [ ]  TODO: check MFDv5 code/documentation to confirm we now actually _always_ match its behaviour, and I haven't left out weird corner cases.

I think the relevant reference code is once again [XLALMakeTimestamps](
https://lscsoft.docs.ligo.org/lalsuite/lalpulsar/_s_f_tutils_8c_source.html#l00412). It has an extra `fudge_down` in it's `ceil` call and also an extra correction for overlapping SFTs which we might need to do as well...

File: pyfstat/make_sfts.py [hunk 1]
```diff
@@ -97,9 +97,10 @@ def __init__(
             frequency evolution and amplitude parameters for injection
             if h0==None or h0==0, these are all ignored
             if h0>0, then Alpha, Delta, cosi need to be set explicitly
-        Tsft: float
-            the sft duration
-        noiseSFTs: str
+        Tsft: int
+            The SFT duration in seconds.
+            Will be ignored if noiseSFTs are given.
```
Comment (Rodrigo-Tenorio) [inline pyfstat/make_sfts.py:102]: But a warning will be thrown if different than the one in the SFTs...?

File: pyfstat/make_sfts.py [hunk 2]
```diff
@@ -584,7 +589,13 @@ def run_makefakedata(self):
         check_ok = self.check_cached_data_okay_to_use(cl_mfd)
         if check_ok is False:
             helper_functions.run_commandline(cl_mfd)
-            logging.info("Successfully wrote SFTs to: {}".format(self.sftfilepath))
+            if os.path.isfile(self.sftfilepath):
+                logging.info(f"Successfully wrote SFTs to: {self.sftfilepath}")
+            else:
+                logging.warning(
```
Comment (Rodrigo-Tenorio) [inline pyfstat/make_sfts.py:595]: This is bad: You may think you are doing stuff on a certain set of SFTs, but maybe you just made a wrong pattern and you are using everything but your actual SFTs. I'd kill the program if this fails.
```suggestion
                logging.critical(
```

File: pyfstat/make_sfts.py [hunk 3]
```diff
@@ -154,16 +155,20 @@ def _get_setup_from_tstart_duration(self):
         self.duration = int(self.duration)
 
         IFOs = self.detectors.split(",")
-        numSFTs = len(IFOs) * [int(float(self.duration) / self.Tsft)]
+        # when duration is not a multiple of Tsft, to match MFDv5
+        # we need to round the number *up*
+        # and also include the overlapping bit at the end in the duration
+        numSFTs = int(np.ceil(float(self.duration) / self.Tsft))  # per IFO
+        effective_duration = numSFTs * self.Tsft
```
Comment (Rodrigo-Tenorio) [inline pyfstat/make_sfts.py:162]: I'd say go for XLALMakeTimestamps .

**Review by Rodrigo-Tenorio — CHANGES_REQUESTED** [review]

File: pyfstat/make_sfts.py (hunk 1 above)
Comment (dbkeitel) [reply pyfstat/make_sfts.py:102]: Yes, or an error if inconsistent between IFOs.

File: pyfstat/make_sfts.py (hunk 2 above)
Comment (dbkeitel) [reply pyfstat/make_sfts.py:595]: I was thinking "MFD has exited successfully, so technically the Writer has succeeded. We warn the user they didn't get back what they thought they would, and then they can decide to abort."

But of course there isn't any real exit status logic here, and manually checking log output isn't an option if the whole workflow is coded up as a script, like in our exampels. So probably you're right. If users want to continue after this, they can still do "try/except" and manually check if any sort of SFT file has been produced.

Comment (Rodrigo-Tenorio) [reply pyfstat/make_sfts.py:595]: A: "Hi, I'd like some water"
B: "Sorry, water machine broke, here's an empty glass"
A: "Thank you, this will quench my thirst".
Then A dies from dehydration.

Comment (dbkeitel) [reply pyfstat/make_sfts.py]: For the sake of an argument, I think in that example the machine wasn't obviously broken, the glass wasn't empty but contained whiskey instead of water, it had a matching "40% alc" label on it, and A only got a funny feeling because they didn't check the label before chugging it.

But in fact you still win this, because turning the warning into an error exposed a failing test and missing multi-IFO-pattern support. Fixed now.

File: pyfstat/make_sfts.py (hunk 3 above)
Comment (dbkeitel) [reply pyfstat/make_sfts.py:162]: WIll make a separate issue for that; we don't support Toverlap so far so the only thing we would gain for now would be to be consistent with lal's fudging in the ceil call, which should only make a difference in rare corner cases.

**Review by Rodrigo-Tenorio — APPROVED** [review]
